src/index.js

```diff
@@ -0,0 +1,360 @@
+    this.workerRegistrationKey = `${this.keyPrefix}:workerIds`;
+    this.assignmentsByConnStr = {};
+    this.assignmentsByWorkerId = {};
+
+    if (!Number.isInteger(this.redundancy) || this.redundancy > 10) {
+      throw new Error(`Specified redundancy '${this.redundancy}' should be an integer less than 10.`);
+    }
+  }
+
+  start() {
+    if (!Array.isArray(this.connectionStrings) || !this.connectionStrings.length) {
+      throw new Error('There are no connectionStrings.');
+    }
+    return Promise.resolve()
+      .then(() => this.registerWorker())
+      .then(() => this.acquireMasterLock())
+      .then(() => this.startPolling());
+  }
+
+  startPolling() {
+    // try to acquire the master lock every 15-30 seconds by default
+    const masterMs = this.masterDuration * 500;
+    this.masterInterval = setInterval(() => {
+      const randomDelay = Math.floor(masterMs * Math.random());
+      setTimeout(() => this.acquireMasterLock(), randomDelay);
+    }, masterMs);
+
+    // re-register this worker every 10 seconds by default
+    const healthcheckMs = this.healthcheckDuration * 1000;
+    this.healthcheckInterval = setInterval(() => this.registerWorker(), healthcheckMs);
+  }
+
+  /**
+   * If the "master" lock is acquired, this worker will be responsible for
+   * assigning oplogs to workers
+   */
+  acquireMasterLock() {
+    const key = `${this.keyPrefix}:master`;
+    return this.lock.acquire(key, {
+      timeout: this.masterDuration * 1000,
+      retries: 0
+    })
+      .then(() => {
+        debug(`${this.workerId} is master.`);
+        return this.assignWorkers();
+      })
+      // ignore lock acquisition errors
+      .catch(err => {});
+  }
+
+  registerWorker() {
+    const key = this.workerRegistrationKey;
+    const now = Date.now();
+    return this.redisClient.zaddAsync(key, now, this.workerId)
+      .then(() => this.startNewAssignments())
+  }
+
+  getWorkerIds() {
+    const expiredTime = Date.now() - (this.healthcheckDuration * 1000);
+    const key = this.workerRegistrationKey;
+    return Promise.resolve()
+      // delete workers that have failed to renew their registration
+      .then(() => this.redisClient.zremrangebyscoreAsync(key, '-inf', expiredTime))
+      // get active workers
+      .then(() => this.redisClient.zrangeAsync(key, 0, -1));
+  }
+
+  getWorkerAssignmentsKey(workerId) {
+    return `${this.keyPrefix}:worker:${workerId}`;
+  }
+
+  readWorkerAssignmentsFromRedis(workerId) {
+    const key = this.getWorkerAssignmentsKey(workerId);
+    return this.redisClient.smembersAsync(key);
+  }
+
+  writeWorkerAssignmentsToRedis() {
+    return Promise.map(Object.keys(this.assignmentsByWorkerId), workerId => {
+      const key = this.getWorkerAssignmentsKey(workerId);
+      const connStrs = this.assignmentsByWorkerId[workerId];
+      return this.redisClient.saddAsync(key, connStrs)
+        .then(() => this.redisClient.expireAsync(key, this.masterDuration * 60));
```

Comment:
Do we need the `masterDuration` here? Is the expiration proportional to the `masterDuration`? If not, should it be a configurable property?